Storage-models.md

```diff
@@ -0,0 +1,393 @@
+* [Common procedures](#common-procedures)
+  * [Create a new model class](#create-a-new-model-class)
+  * [Add a pseudonymizable model](#add-a-pseudonymizable-model)
+  * [Modify a model field](#modify-a-model-field)
+    * [Add an index to a field](#add-an-index-to-a-field)
+    * [Remove an old field from a model](#remove-an-old-field-from-a-model)
+* [Contact](#contact)
+
+## Introduction
+
+You might be used to programs holding data in variables, but for an application like Oppia, we need some way to store that data such that it survives the application shutting down. We don't want to lose all our data just because we release a new version of Oppia! To store data persistently, you might think about storing it in a file like this:
+
+```text
+ User        Application               File system
+ ----        -----------               -----------
+  |  GET /prefs   |                         |
+  | ------------> |                         |
+  |               |   f = open('f.json')    |
+  |               |   data = json.load(f)   |
+  |               | ----------------------> |
+  |               |                         |
+  |               |          data           |
+  |               | <---------------------- |
+  | data['prefs'] |                         |
+  | <------------ |                         |
+```
+
+This works, but it turns out we can do better by using a dedicated database server. At Oppia, we use Google Cloud's [Datastore](https://cloud.google.com/datastore) like this:
+
+```text
+ User        Application                Datastore
+ ----        -----------                ---------
+  |  GET /prefs   |                         |
+  | ------------> |                         |
+  |               |  Retrieve preferences   |
+  |               | ----------------------> |
+  |               |                         |
+  |               |          prefs          |
+  |               | <---------------------- |
+  |      prefs    |                         |
+  | <------------ |                         |
+```
+
+To interact with the datastore, we use Google's [ndb library](https://googleapis.dev/python/python-ndb/latest/index.html). We refer to the code that uses ndb and defines what kind of data we store in the datastore as the "storage layer." In this guide, we'll discuss how the storage layer works and how to make some common storage layer changes.
+
+## How the storage layer works
+
+### Storage model concepts
+
+We organize our data using _models_, which you can think of like Python classes. A model describes an object that can have properties of various types. Instances of these models can then be created with particular values for those properties. However, unlike Python classes, a model has only data (no functions), and a model's properties are strictly typed.
+
+For example, we might define a user model with the following properties:
+
+* `id`: An integer
+* `username`: A string
+* `admin`: A boolean
```

Comment:
Perhaps this should be more like user_is_admin (this is kind of our convention for naming booleans in Oppia -- the names should obviously be storing true/false).

Also just FYI, we don't tend to go with the alternative is_admin because the "is" at the beginning suggests a function rather than a variable. (I.e. sticking to the rule of "functions/methods start with verbs, variables don't".)